File: leetcode/cpp/205. 同构字符串.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if (s.size() != t.size()) {
            return false;
        }

        std::unordered_map<char, char> d;
        std::unordered_set<char> visited;

        for (int i = 0; i < s.size(); i++) {
            char si = s[i], ti = t[i];
            if (d.count(si) == 0 && visited.count(ti) == 0) {
                d[si] = ti;
                visited.insert(ti);
            } else if (d.count(si) == 0 && visited.count(ti) > 0) {
                return false;
            } else if (d.count(si) > 0 && d[si] != ti) {
                return false;
            }
        }
        return true;
    }

	// ===== 更加高效的方法 =====
	// 判断两字符串对应位置的字符，在第一次出现的位置是否相同
	bool isIsomorphic(string s, string t) {
        if (s.size() != t.size()) {
            return false;
        }
        vector<int> s_first_index(256, -1), t_first_index(256, -1);
        for (int i = 0; i < s.size(); i++) {
            if (s_first_index[s[i]] != t_first_index[t[i]]) {
                return false;
            }
            s_first_index[s[i]] = t_first_index[t[i]] = i;
        }
        return true;
    }
};

File: leetcode/python/205. 同构字符串.py
#!/bin/bash python
"""
205. 同构字符串

思路：
hash
"""

class Solution:
    # ===== 第一版 =====
    def isIsomorphic(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        if len(s) != len(t):
            return False

        d = dict()      # 记录由s到t的映射
        visited = set() # 记录t中已访问过的元素

        for i in range(len(s)):
            si, ti = s[i], t[i]
            if si not in d:
                if ti in visited:
                    return False
                else:
                    d[si] = ti 
                    visited.add(ti)
            else:
                if d[si] != ti:
                    return False
        return True

    # ===== 第二版 =====
    def isIsomorphic2(self, s, t):
        """
        需要两个hash
        """
        ss, tt = dict(), set()
        for i in range(len(s)):
            si, ti = s[i], t[i]
            if si not in ss and ti not in tt:
                ss[si] = ti
                tt.add(ti)
            elif si not in ss and ti in tt:
                return False
            elif si in ss and ti != ss[si]:
                return False
        return True